ui/assets/js/signal.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Signal represents a collection of observers that can be notified with
+ * parameterized event dispatch. Listeners can tap into a signal to
+ * receive callbacks when a signal is raised.
+ */
+export class Signal {
+	constructor() {
+		this.listeners = [];
+	}
+
+	/**
+	 * Begin listening to this signal with the given callback, l.
+	 *
+	 * @param {function} l
+	 */
+	tap(l) {
+		// Make a copy of the listeners to avoid the all too common
+		// subscribe-during-dispatch problem
+		this.listeners = this.listeners.slice(0);
+		this.listeners.push(l);
```

Comment:
You could do something like this in here (although I understand if you'd rather leave this as-is): 
```suggestion
		this.listeners = [...this.listeners, l]
```